lib/collection/src/shards/local_shard/query.rs

```diff
@@ -147,45 +147,44 @@ impl LocalShard {
-                let search_requests = sources
+                // create single search request for rescoring query
+                let point_ids = sources
                     .into_iter()
-                    .map(|source| {
-                        // extract target point ids to score
-                        let point_ids: HashSet<_> =
-                            source.iter().map(|scored_point| scored_point.id).collect();
-                        // create filter for target point ids
-                        let filter = Filter::new_must(segment::types::Condition::HasId(
-                            HasIdCondition::from(point_ids),
-                        ));
-                        CoreSearchRequest {
-                            query: query_enum.clone(),
-                            filter: Some(filter),
-                            params: None,
-                            limit,
-                            offset: 0,
-                            with_payload: None, // the payload has already been fetched
-                            with_vector: None,  // the vector has already been fetched
-                            score_threshold: None,
-                        }
-                    })
-                    .collect();
+                    .flatten()
+                    .map(|point| point.id)
+                    .collect::<HashSet<_>>();
+
+                // create filter for target point ids
+                let filter = Filter::new_must(segment::types::Condition::HasId(
+                    HasIdCondition::from(point_ids),
+                ));
+
+                let search_request = CoreSearchRequest {
+                    query: query_enum,
+                    filter: Some(filter),
+                    params: None,
+                    limit,
+                    offset: 0,
+                    with_payload: None, // the payload is fetched separately
+                    with_vector: None,  // the vector is fetched separately
+                    score_threshold: None,
+                };
+
                 let rescoring_core_search_request = CoreSearchRequestBatch {
-                    searches: search_requests,
+                    searches: vec![search_request],
                 };
-                let core_results = self
+
+                let top = self
                     .do_search(
                         Arc::new(rescoring_core_search_request),
                         search_runtime_handle,
                         timeout,
                     )
-                    .await?;
-                let top = core_results
-                    .into_iter()
-                    .flatten()
-                    .sorted_unstable()
-                    .take(limit)
-                    .collect();
+                    .await?
+                    // One search request is sent. We expect only one result
+                    .pop()
+                    .unwrap_or_default();
```

Comment:
I wonder if panicing with a descriptive error may be better here :thinking:

Reply:
Maybe not panic but logging?

Reply:
Better.

The main point was, just taking the default here if something goes wrong may create unexpected problems down the line that are hard to debug.

Reply:
It now returns a service error